doc/stdenv/stdenv.chapter.md

```diff
@@ -362,6 +364,42 @@ Unless set to `false`, some build systems with good support for parallel buildin
 
 ### Special variables {#special-variables}
 
+#### `__cleanAttrs` {#var-__cleanAttrs}
+
+By default, `mkDerivation` will expose its arguments in the returned package attribute set. This is unnecessary and leads to some confusion and doubt.
+
+New packages may pass `__cleanAttrs = true;` to `mkDerivation`, so that it will return a minimal set of package attributes, which package authors can extend via [`passthru`](#var-stdenv-passthru).
+
+Existing packages may be modified to pass `__cleanAttrs = "warn";`, so that the legacy attributes remain available, but they will print a helpful warning when they are accessed. Doing this is best avoided until the packaging function used supports recursively defined arguments like [`mkDerivation` does](#mkderivation-recursive-attributes).
+
+Benefits of `__cleanAttrs`:
+
+ - Users don't have to sift through unnecessary attributes when exploring a package in the `nix repl`.
+ - Users can confidently use non-standard attributes that a package provides.
+ - Package authors know which variables are intended for use by the builder. They can change the builder environment with confidence that they don't break consumers of their package.
+ - It is a little bit more efficient in terms of CPU cycles and memory use.
+
+##### What to do when warned {#warning-package-attr-impl-detail}
+
+If you encounter the warning `The attribute ... of package ... is an implementation detail`, you are invited to help us explicitly support ways in which a package attribute set may used, so that Nixpkgs contributors and users will be aware of your use case.
```

Comment:
This could be phrased to be more immediately actionable.

```suggestion
If you encounter the warning `The attribute ... of package ... is an implementation detail`, please make Nixpkgs contributors and users aware of your particular use case, and help providing explicit ways to use package attribute sets, by opening an [issue](https://github.com/NixOS/nixpkgs/issues) or (https://github.com/NixOS/nixpkgs/pulls).
```